core/controllers/editor_test.py

```diff
@@ -1381,12 +1392,86 @@ def test_for_assign_role_for_exploration(self):
         self.assert_cannot_voiceover(exp_id)
         self.logout()
 
+        # Check that COLLABORATOR 4, who cannot edit but can voiceover,
+        # cannot add a new State 6.
+        self.login(self.COLLABORATOR4_EMAIL)
+        csrf_token = self.get_new_csrf_token()
+
+        reader_dict = self.get_json(
+            '%s/%s' % (feconf.EXPLORATION_DATA_PREFIX, exp_id))
+        self.assertNotIn('State 4', reader_dict['states'])
+
+        # Since check_can_edit_activity always return False, therefore
+        # user will behave as a voiceover artist.
+        get_voiceover_swap = self.swap_to_always_return(
+            rights_manager, 'check_can_edit_activity', value=False)
+
+        with get_voiceover_swap:
+            self.assert_cannot_edit(exp_id)
+            self.assert_can_voiceover(exp_id)
+
+            self.put_json(
+                '%s/%s' % (feconf.EXPLORATION_DATA_PREFIX, exp_id),
+                {
+                    'version': exploration.version,
+                    'commit_message': 'Added State 4',
+                    'change_list': [{
+                        'cmd': 'add_state',
+                        'state_name': 'State 4'
+                    }]
+                },
+                csrf_token=csrf_token,
+                expected_status_int=500
+            )
+
+            reader_dict = self.get_json(
+                '%s/%s' % (feconf.EXPLORATION_DATA_PREFIX, exp_id))
+            self.assertNotIn('State 4', reader_dict['states'])
+            self.assertNotIn(
+                self.COLLABORATOR3_USERNAME,
+                reader_dict['rights']['editor_names']
+            )
+
+            # Check that collaborator 4 cannot add new members.
+            exploration = exp_fetchers.get_exploration_by_id(exp_id)
+            rights_url = '%s/%s' % (feconf.EXPLORATION_RIGHTS_PREFIX, exp_id)
+            response = self.put_json(
+                rights_url, {
+                    'version': exploration.version,
+                    'new_member_username': self.COLLABORATOR3_USERNAME,
+                    'new_member_role': rights_domain.ROLE_EDITOR,
+                },
+                csrf_token=csrf_token,
+                expected_status_int=401
+            )
+            error_msg = (
+                'You do not have credentials to change rights '
+                'for this exploration.'
+            )
+            self.assertEqual(response['error'], error_msg)
+            reader_dict = self.get_json(
+                '%s/%s' % (feconf.EXPLORATION_DATA_PREFIX, exp_id))
+            self.assertNotIn(
+                self.COLLABORATOR3_USERNAME,
+                reader_dict['rights']['editor_names']
+            )
```

Comment:
I think you're still missing a pre-check for this. Remember that seanlip's review said "Here and below," so make sure you also update the similar situations "below"

Reply:
@U8NWXD I had already added a pre-check for this. 
Refer: lines 1430-1433
![image](https://user-images.githubusercontent.com/75239620/151326107-20530e8f-62be-49d6-a042-4a15c5aa0cbd.png)
If I need to do something else then please let me know.

Reply:
I think the problem is that you're mixing up the end of the previous "test" and the start of this "test". For example, if we regard the block above as the setup for this part of the test, then it seems weird to check whether State 4 has been added since it has nothing to do with this test.

To remedy this I suggest you split this long test into multiple shorter tests and give them individual names that are clear. Each test should have a clear setup / pre-check / action / post-check -- and that's it. If two tests are strongly related, you might want to combine them in a single test method, but try not to do that if they aren't. That will help improve the clarity of this set of tests.

Reply:
Note: this guidance mirrors that from the page I sent you before, see points 1 and 2: https://github.com/oppia/oppia/wiki/Backend-tests#guidelines-for-writing-good-tests

Reply:
@seanlip I understand your concern. I also found the testing style to be quite absurd. So, I have split the whole test into 6 different tests now. Kindly, review it.